### PR TITLE
Point at `:` when using it instead of `;`

### LINKED ISSUE #42057: Better error message than E0573 when you typo a colon in place of a semicolon
I made this typo in some code I was writing just now, a colon in place of a semicolon after a function call, with another function call on the next line:
```rust
fn f() {}

fn main() {
    f():
    f();
}
```

[Run it on play](https://play.rust-lang.org/?gist=8dcb2993683495112686576d6766edfc&version=stable&backtrace=0).

The compiler errors with:
```
rustc 1.17.0 (56124baa9 2017-04-24)
error[E0573]: expected type, found function `f`
 --> <anon>:5:5
  |
5 |     f();
  |     ^^^ not a type
```

This is technically correct, but it was confusing and it took me a few minutes to realize what I had done. It would be nicer if we could guess in some circumstances that the user probably meant to put a semicolon there and say something to that effect.

### PR DESCRIPTION
**Pull request by estebank**
When triggering type ascription in such a way that we can infer a
statement end was intended, add a suggestion for the change. Always
point out the reason for the expectation of a type is due to type
ascription.

Fix #42057, #41928.

### CODE REVIEW

File: src/test/ui/suggestions/type-ascription-instead-of-statement-end.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(type_ascription)]
+
+fn main() {
+    println!("test"):
+    0;
```
Comment (KalitaAlexey) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.rs]: Will it work if a colon was here?

File: src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr [hunk 2]
```diff
@@ -0,0 +1,13 @@
+error: expected type, found `0`
+  --> $DIR/type-ascription-instead-of-statement-end.rs:15:5
+   |
+14 |     println!("test"):
+   |                     -
+   |                     |
+   |                     help: did you mean to end the statement here instead? `;`
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr]: this sort of looks funny, perhaps because of the `^` having no label. I think I'd prefer:

```
error: expected type, found `0`
  --> $DIR/type-ascription-instead-of-statement-end.rs:15:5
   |
14 |     println!("test"):
   |                     -
   |                     |
   |                     help: did you mean to use a `;` here?
15 |     0;
   |     ^ expecting a type here because of type ascription

error: aborting due to previous error
```

Comment (estebank) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr]: Done.

File: src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr [hunk 3]
```diff
@@ -0,0 +1,16 @@
+error: expected type, found `0`
+  --> $DIR/type-ascription-instead-of-statement-end.rs:15:5
+   |
+14 |     println!("test"):
+   |                     - help: did you mean to end the statement here instead? `;`
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr]: Huh, this looks strange to me somehow. Why not say "help: did you mean to use `;` here?"? I think users might not know that "end the statement" means `;` -- but also, I am not sure about this "floating" `;` that appears at the end of the sentence.

Comment (estebank) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr]: The presentation is because it is an inline suggestion. Should I change the code so that we can control wether the code to be replaced should be displayed when inline?

Comment (nikomatsakis) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr]: Hmm; I think either change the code, or change the wording. But yeah probably just having the option to "suppress" the suggestion when displayed inline seems good.

**Comment by shepmaster** [comment]
This is your polite 7-day ping @estebank — when do you think you'll be able to respond to the comments?

File: src/test/ui/suggestions/type-ascription-instead-of-statement-end.rs (hunk 1 above)
Comment (estebank) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.rs]: The output in that case is

```rust
error: expected type, found `0`
  --> ../../src/test/ui/suggestions/type-ascription-instead-of-statement-end.rs:15:5
   |
14 |     println!("test"):
   |                     - help: did you mean to use `;` here instead?
15 |     0:
   |     ^ expecting a type here because of type ascription
```

File: src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr (hunk 3 above)
Comment (estebank) [inline src/test/ui/suggestions/type-ascription-instead-of-statement-end.stderr]: Done.

**Comment by nikomatsakis** [comment]
Failure:

```
[01:08:34] failures:
[01:08:34]     [ui] ui/issue-22644.rs
```

The new message looks improved, though. =)

**Comment by estebank** [comment]
@nikomatsakis fixed

**Comment by estebank** [comment]
@nikomatsakis covered another case that fails during `resolve` instead of `parse`:

```rust
fn f() {}

fn main() {
    f():
    f();
}
```

**Comment by estebank** [comment]
@nikomatsakis ping

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
cc @oli-obk, I don't this suggestion follows the guidelines, but I also think it looks good. Perhaps we need to adjust the guidelines to account for the "suppress example code" option?

**Comment by oli-obk** [comment]
I'm assuming you mean that the guildelines aren't followed by the use of `did you mean`. This guideline exists because of https://github.com/rust-lang/rust/pull/42033#discussion_r117135856 which notes that the `did you mean` sounds condescending. I don't have a strong opinion on this guideline though.

"Try using a `;` instead:" looks fine to me. Otherwise let's adjust the guidelines if that's desired.

**Comment by luser** [comment]
@estebank Thanks for making this error message better!

**Comment by Nokel81** [comment]
Which error message got finalized?

**Comment by estebank** [comment]
@Nokel81 sorry, I don't understand the question. Could you rephrase? This PR adds a label stating "help: did you mean to use `;` here?" when type ascription (triggered by `:`) ends the previous line to the type error, as it probably was a mistyped `:`

**Comment by Nokel81** [comment]
I was talking about the guideline of not using `did you mean` that was referenced several times near the end of this PR chain. Was the guidelines followed or was the original error message kept as `help: did you mean to use ; here instead?`

**Comment by estebank** [comment]
@Nokel81 ah, I see. No, I didn't change the message before it got merged. I can post a PR later today/tomorrow changing the wording, if you want.

**Comment by Nokel81** [comment]
No it is fine, I personally feel that there are some cases where `did you mean` makes sense. Mostly for small syntactic typos like this
